nemo/core/classes/optimizers.py

```diff
@@ -0,0 +1,281 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from argparse import ArgumentParser
+from functools import partial
+
+import torch
+import torch.optim as optim
+from torch.optim import adadelta, adagrad, adamax, rmsprop, rprop
+from torch.optim.optimizer import Optimizer
+
+__all__ = ['Novograd', 'get_optimizer', 'register_optimizer', 'parse_optimizer_args', 'add_optimizer_args']
+
+
+AVAILABLE_OPTIMIZERS = {
+    'sgd': optim.SGD,
+    'adam': optim.Adam,
+    'adamw': optim.AdamW,
+    'adadelta': adadelta.Adadelta,
+    'adamax': adamax.Adamax,
+    'adagrad': adagrad.Adagrad,
+    'rmsprop': rmsprop.RMSprop,
+    'rprop': rprop.Rprop,
+}
+
+
+def _boolify(s):
+    if s == 'True' or s == 'true':
+        return True
+    if s == 'False' or s == 'false':
+        return False
+    raise ValueError('Not Boolean Value!')
+
+
+def _autocast(value):
+    # if value is itself None, dont parse
+    if value is None:
+        return None
+
+    # If value is comma seperated list of items, recursively parse all items in list.
+    if "," in value:
+        values = value.split(',')
+        values = [_autocast(value) for value in values]
+        return values
+
+    # If value is string `none` or `None`, parse as None
+    if value == 'none' or 'None':
+        return None
+
+    # Try type cast and return
+    for cast_type in (int, float, _boolify):
+        try:
+            return cast_type(value)
+        except Exception:
```

Comment:
shouldn't we fail in this case?

Reply:
No, it will try all possibly casts - in the order of dtype - int then float then bool. Above it tests for None. If it escapes everything, its a string type value.